README
NAME
    Device::PCA9685 - I2C interface to PCA9685 16 channel 12 bit PWM
    Generator

VERSION
    version 0.002

SYNOPSIS
         use Device::PWMGenerator::PCA9685;
 
         my $dev = Device::PWMGenerator::PCA9685->new(
             I2CBusDevicePath => '/dev/i2c-1', # this would be '//dev/i2c-dev-0 for Model A Pi
             debug            => 1,
             frequency        => 400, #Hz
         );
         $dev->enable();
         $dev->setChannelPWM(4,0,$dutycycle); # Duty cycle values between 0 and 4096 channel 4

DESCRIPTION
    This module provides a Object oriented interface to Adafruit 16 channel
    12 bit PWM Generator(PCA9685). Before you use this module make sure you
    follow install Prerequisites for Device::SMBus

ATTRIBUTES
  I2CBusDevicePath
    this is the device file path for your I2CBus that the PCA9685 is
    connected on e.g. /dev/i2c-1 This must be provided during object
    creation.

  PWMGenerator
        $self->PWMGenerator->enable();
        $self->PWMGenerator->setChannelPWM($channel,$pulseOnPoint,$pulseOffPoint);

    The PCA9685 offers a 12 bit resolution which means across a duty cycle
    you may set and unset the PWM at 4096 different point. Range of values
    for $pulseOnPoint and $pulseOffPoint is 0 to 4095 Range of values of
    $channel is 0 to 15

    This is a object of Device::PWMGenerator::PCA9685

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through github at
    <https://github.com/shantanubhadoria/device-pca9685/issues>. You will be
    notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/shantanubhadoria/device-pca9685>

      git clone git://github.com/shantanubhadoria/device-pca9685.git

AUTHOR
    Shantanu Bhadoria <shantanu at cpan dott org>

CONTRIBUTORS
    *   Shantanu <shantanu@cpan.org>

    *   Shantanu Bhadoria <shantanu@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Shantanu Bhadoria.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changelog for Device-PCA9685

0.002 2013-12-15T02:36:26
 - Fixed pod spellings [Shantanu]
 - Fixed POD [Shantanu]

0.001 2013-09-27T11:07:52
 - v0.001
   
    - Possibly first release [Shantanu]
    - Refactoring Submodule to allow easy loading of similar subdevices as
      plugins using a common prefix in MOP [Shantanu Bhadoria]
    - First Commit [Shantanu Bhadoria] [Shantanu]
 - Possibly first release [Shantanu]
 - Refactoring Submodule to allow easy loading of similar subdevices as
   plugins using a common prefix in MOP [Shantanu Bhadoria]
 - First Commit [Shantanu Bhadoria]



lib/Device/PWMGenerator/PCA9685.pm
use strict;
use warnings;

package Device::PWMGenerator::PCA9685;

# PODNAME: Device::PWMGenerator::PCA9685
# ABSTRACT: I2C interface to PWM Generator on PCA9685 using Device::SMBus
#
# This file is part of Device-PCA9685
#
# This software is copyright (c) 2013 by Shantanu Bhadoria.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
our $VERSION = '0.002'; # VERSION

# Dependencies
use 5.010;
use Moose;
use POSIX;

use Time::HiRes qw(usleep);

extends 'Device::SMBus';


use constant {
    MODE1 => 0x00,

    SUBADR1 => 0x02,
    SUBADR2 => 0x03,
    SUBADR3 => 0x04,

    PRESCALE => 0xFE,

    LED0_ON_L  => 0x06,
    LED0_ON_H  => 0x07,
    LED0_OFF_L => 0x08,
    LED0_OFF_H => 0x09,
};


has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x40,
);


has frequency => (
    is       => 'rw',
    required => 1,
    trigger  => \&_frequencySet,
);

has debug => (
    is      => 'ro',
    default => 0,
);

sub _frequencySet {
    my ( $self, $newFrequency, $oldFrequency ) = @_;

    # PCA9685 allows you to set the frequency using a register it calls PRESCALE
    say "Making PCA9685 go to sleep" if $self->debug;
    my $oldMode = $self->readByteData(MODE1);
    my $newMode = ( $oldMode & 0x7f ) | 0x10
      ; #set sleep bit while preserving other bits and ensuring that restart bit is set to 0 when writing to register
    $self->writeByteData( MODE1, $newMode );

    # Calculate Prescale from Frequency. Clock is 25 MHz on PCA8695
    my $prescale = floor( ( 25000000 / ( 4096 * $newFrequency ) ) + 0.5 ) - 1;
    say
"Setting frequency to $newFrequency Hz. Calcualting prescale value \nprescale=round( 25000000/(4096 * frequency)) - 1 \nfrom frequency as $prescale(PCA9685 clock is at 25MHz or 25000000 Hz)"
      if $self->debug;
    if ( $prescale >= 0x03 ) {
        $self->writeByteData( PRESCALE, $prescale );
    }
    else {
        die
"Bad frequency value. Can't set prescale less than 3. reduce your frequency value";
    }
    $self->writeByteData( MODE1, $oldMode );
    usleep(5000);
    $self->writeByteData( MODE1, $oldMode | 0x80 );    # set extclk bit
}


sub enable {
    my ($self) = @_;
    say "Setting up PCA9685" if $self->debug;
    $self->writeByteData( MODE1, 0b00000000 );
}


sub setChannelPWM {
    my ( $self, $channel, $on, $off ) = @_;

    say
"Setting Channel $channel PWM on at $on step, off at $off step in 0 to 4095 steps at "
      . $self->frequency . "Hz"
      if $self->debug;
    $self->writeByteData( LED0_ON_L + ( 4 * $channel ), $on & 0xff );
    $self->writeByteData( LED0_ON_H + ( 4 * $channel ), $on >> 8 );

    $self->writeByteData( LED0_OFF_L + ( 4 * $channel ), $off & 0xff );
    $self->writeByteData( LED0_OFF_H + ( 4 * $channel ), $off >> 8 );
}

1;

__END__

=pod

=head1 NAME

Device::PWMGenerator::PCA9685 - I2C interface to PWM Generator on PCA9685 using Device::SMBus

=head1 VERSION

version 0.002

=head1 ATTRIBUTES

=head2 I2CDeviceAddress

Contains the I2CDevice Address for the bus on which your PWM Generator is connected. It would look like 0x40. Default value is 0x40, If you have not set any of the six jumpers on the PCA9685, this should be the correct address and you will not have to modify this value in most cases.

=head2 Frequency

Frequency of the PWM Pulse in Hz

=head1 METHODS

=head2 enable

    $self->enable()

This function is just a placeholder. It is not required to call this function for this particular chipset

=head2 setChannelPWM

    $self->setChannelPWM($channel,$pulseOnPoint,$pulseOffPoint)
    The PCA9685 offers a 12 bit resolution which means across a duty cycle you may set and unset the PWM at 4096 different point. 
    Range of values for $pulseOnPoint and $pulseOffPoint is 0 to 4095
    Range of values of $channel is 0 to 15

=head1 REGISTERS

=head2 MODE1

=head2 SUBADR1

=head2 SUBADR2

=head2 SUBADR3

=head2 PRESCALE

=head2 LED0_ON_L

=head2 LED0_ON_H

=head2 LED0_OFF_L

=head2 LED0_OFF_H

=head2 LED0_OFF_H

=head1 AUTHOR

Shantanu Bhadoria <shantanu at cpan dott org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Shantanu Bhadoria.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This Makefile.PL for  was generated by Dist::Zilla::Plugin::MakeMaker::Awesome 0.15.
# Don't edit it but the dist.ini used to construct it.

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "I2C interface to PCA9685 16 channel 12 bit PWM Generator ",
  "AUTHOR" => "Shantanu Bhadoria <shantanu at cpan dott org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Device-PCA9685",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Device::PCA9685",
  "PREREQ_PM" => {
    "Device::SMBus" => 0,
    "Moose" => 0,
    "POSIX" => 0,
    "Time::HiRes" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



